builder/builder.go

```diff
@@ -114,8 +115,9 @@ func NewBuilder(ctx context.Context, opt Opt) (*Builder, error) {
 			enttlmnts:       opt.Enttlmnts,
 			saveInlineCache: opt.SaveInlineCache,
 		},
-		opt:      opt,
-		resolver: nil, // initialized below
+		opt:         opt,
+		pullPingMap: gatewaycrafter.NewPullPingMap(),
```

Comment:
Could this be one per-build (basically, initialized in `convertAndBuild`) rather than just just global for the entire `Builder`?

Reply:
Since the map is so tied conceptually to outputting via the local registry, I wonder if it would be nicer to just tuck this map within the gateway crafter object itself.

The global `onPull` function could be provided during the construction of the crafter (which currently takes place in `convertAndBuild`).

Reply:
> Could this be one per-build (basically, initialized in convertAndBuild) rather than just just global for the entire Builder?

That could work!

> Since the map is so tied conceptually to outputting via the local registry, I wonder if it would be nicer to just tuck this map within the gateway crafter object itself.

This is harder because multiple gatewaycrafter instances are created per session (one for each wait/end block, plus for the final return of the build function), where as the pullpingmap is only created a single time due to there only being a single onPull function created once per session.

If instead there was a way to tell buildkit to call back a different onPull function, we could reference the per-export specific instance of the gatewaycrafter in
https://github.com/earthly/earthly/blob/1c5ebbd5e7d2bcd710352dc51536af988cdc3028/earthfile2llb/wait_block.go#L122
```
        otherOnPullFunc := func(...) {
            gwCrafter.PullPingMap.Get(...)
            ...etc...
        }
	refs, metadata := gwCrafter.GetRefsAndMetadata()
	err := gatewayClient.Export(ctx, gwclient.ExportRequest{
		Refs:     refs,
		Metadata: metadata,
	},
	otherOnPullFunc)
```

but this would require more work in buildkit to allow us to essentially pass a `void*` pointer that the per-session onPull function could use to dispatch back to this otherOnPullFunc.

Reply:
Ah gotcha - so the gateway crafter is one per export, whereas the map and the onPull function is one per convertAndBuild call.